src/ripple/app/tx/impl/InvariantCheck.cpp

```diff
@@ -0,0 +1,196 @@
+
+#include <ripple/app/tx/impl/InvariantCheck.h>
+#include <ripple/basics/Log.h>
+
+namespace ripple {
+
+/**
+ * @brief Invariant: A transaction must not create XRP and should only destroy
+ * XRP, up to the transaction fee.
+ *
+ * For this check, we start with a signed 64-bit integer set to zero. As we go
+ * through the ledger entries, look only at account roots, escrow payments,
+ * and payment channels.  Remove from the total any previous XRP values and add
+ * to the total any new XRP values. The net balance of a payment channel is
+ * computed from two fields (amount and balance) and deletions are ignored
+ * for paychan and escrow because the amount fields have not been adjusted for
+ * those in the case of deletion.
+ *
+ * The final total must be less than or equal to zero and greater than or equal
+ * to the negative of the tx fee.
+ *
+ */
+struct XRPNotCreated :
+    public InvariantCheckBase
+{
+    XRPNotCreated() : drops_(0) {};
+
+    void
+    visitEntry(
+        uint256 const& /*index*/,
+        bool isDelete,
+        std::shared_ptr <SLE const> const& before,
+        std::shared_ptr <SLE const> const& after) override
+    {
+        if(before)
+        {
+            switch (before->getType())
+            {
+            case ltACCOUNT_ROOT:
+                drops_ -= (*before)[sfBalance].xrp().drops();
+                break;
+            case ltPAYCHAN:
+                drops_ -= ((*before)[sfAmount] - (*before)[sfBalance]).xrp().drops();
+                break;
+            case ltESCROW:
+                drops_ -= (*before)[sfAmount].xrp().drops();
+                break;
+            default:
+                break;
+            }
+        }
+
+        if(after)
+        {
+            switch (after->getType())
+            {
+            case ltACCOUNT_ROOT:
+                drops_ += (*after)[sfBalance].xrp().drops();
+                break;
+            case ltPAYCHAN:
+                if (! isDelete)
+                    drops_ += ((*after)[sfAmount] - (*after)[sfBalance]).xrp().drops();
+                break;
+            case ltESCROW:
+                if (! isDelete)
+                    drops_ += (*after)[sfAmount].xrp().drops();
+                break;
+            default:
+                break;
+            }
+        }
+    };
+
+    bool
+    finalize(STTx const& tx, TER /*tec*/, beast::Journal const& j) override
+    {
+        auto fee = tx.getFieldAmount(sfFee).xrp().drops();
+        if(-1*fee <= drops_ && drops_ <= 0)
+        {
+            return true;
+        }
+        else
+        {
+            JLOG(j.error()) << "Invariant failed: XRP net change was "
+                << drops_ << " on a fee of " << fee;
+            return false;
+        }
+    };
+
+private:
+    int64_t drops_;
+};
+
+
+/**
+ * @brief Invariant: we cannot remove an account ledger entry
+ *
+ * an account root should never be the target of a delete
+ */
+struct AccountRootsNotDeleted :
+    public InvariantCheckBase
```

Comment:
Note: `structs` default to public inheritance, so `public` is redundant here. No change needed.